rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -233,15 +233,21 @@ std::shared_ptr<ClientData> ClientData::make(
     return nullptr;
   }
 
-  auto client_data = std::shared_ptr<ClientData>(
-    new ClientData{
-      node,
-      std::move(entity),
-      request_members,
-      response_members,
-      std::move(request_type_support),
-      std::move(response_type_support)
-    });
+  std::lock_guard<std::mutex> lock(num_in_flight_mutex);
+  std::vector<std::shared_ptr<ClientData>> duplicate_pointers;
+  std::shared_ptr<ClientData> client_data;
+  do {
+    client_data = std::shared_ptr<ClientData>(
+      new ClientData{
+        node,
+        std::move(entity),
+        request_members,
+        response_members,
+        std::move(request_type_support),
+        std::move(response_type_support)
+      });
+    duplicate_pointers.push_back(client_data);
```

Comment:
If we didn't do this, then we could conceivably get the same address over and over again.  Consider the case where we have the pointer 0x1 in `deleted_clients`.  The first call to `new ClientData` might return 0x1.  At that point, we'd detect that this is in `deleted_clients`, and loop around again.  But right before we make the second call to `new ClientData`, we drop the reference to the old one, at which point it may be freed.  At that point the allocator might decide to give us address 0x1 again, and we'd never make any progress here.

By holding all of the shared_ptr in a vector like this, we guarantee that the allocator will give us a new address.  And then at the end of the function, when `std::vector` goes out of scope, we'll free all of them.